### PR TITLE
Add native conv1d op and route framework-lowered 1D convolutions to it

### PR DESCRIPTION
**Pull request by sonalibaskaran2499**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/47799)

### Problem description
torch-xla lower nn.Conv1d to a 2D stablehlo.convolution with a size-1 spatial dim, so it previously lowered to ttnn.conv2d. For depthwise 1D convs, the conv2d path with config_tensors_in_dram=true hangs in tt-metal ([#47799](https://github.com/tenstorrent/tt-metal/issues/47799)). Surfaced by the Qwen3.5 gated-delta depthwise causal conv.

### What's changed

- New conv1d op across the stack (TTIR + TTNN defs/verifiers, TTIR→TTNN, EmitC/EmitPy, flatbuffer schema + serialization, runtime execution, TTIRBuilder/golden, tests), modeling ttnn::conv1d.
- Routing: tryEmitDegenerate2dAsConv1d detects a 2D conv with one spatial dim of extent 1 in both input and kernel, squeezes it, and emits ttir.conv1d. Genuine 2D convs are untouched.
- Config: conv1d uses an L1 config (config_tensors_in_dram=false) to avoid the hang; conv2d/conv_transpose2d stay in DRAM.
- Layout workarounds (mirroring conv2d): input forced row-major, weight/bias moved to host row-major - ttnn::conv1d does its own runtime weight-prep and requires a row-major host weight.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by sgligorijevicTT** [comment]
+1 for this change, I too need this op and would appreciate if this PR could make it in sooner rather than later

**Comment by sdjordjevicTT** [comment]
> +1 for this change, I too need this op and would appreciate if this PR could make it in sooner rather than later

Makes sense, but note that the metal underneath actually calls con2d. Still, this makes sense to me as it is better to lower this abstraction as much as possible...

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp [hunk 1]
```diff
@@ -3373,6 +3373,24 @@ SamplingOp::getOpRuntime(const std::vector<TTNNLayoutAttr> &inputs,
...
+//===----------------------------------------------------------------------===//
+// Conv1dOp - TTNN Op Model Interface
+//===----------------------------------------------------------------------===//
+
+llvm::Expected<op_model::OpConstraints>
+Conv1dOp::getOpConstraints(const std::vector<TTNNLayoutAttr> &inputs,
+                           const OpConfig &opConfig) {
+  return issueErrorForGetOpConstraints(
+      getOperation(), detail::ReasonForLackOfSupport::MissingMetalDefinition);
+}
```
Comment (rpavlovicTT) [inline lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp:3425]: I don't get it how are we missing metal definition if you can invoke this op through runtime?

**Review by jgrimTT — APPROVED** [review]

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp (hunk 1 above)
Comment (sonalibaskaran2499) [reply lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp:3425]: MissingMetalDefinition was the wrong enum to use here. It's resolved now I removed the stub, replaced by a real OpModel<Conv1dOp> that invokes ::ttnn::conv1d (same path as the runtime) and reuses the conv2d prepare helpers.

**Review by rpavlovicTT — APPROVED** [review]

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 2]
```diff
@@ -1573,6 +1554,67 @@ mlir::Operation *mlir::tt::ttir::Conv2dOp::rewriteWithQuantizedInputs(
...
+  RankedTensorType weightType = getWeight().getType();
+  RankedTensorType outputType = getResult().getType();
+
+  int64_t inChannels = inputType.getDimSize(getChannelDim());
+  int64_t outChannels = outputType.getDimSize(getChannelDim());
+  uint32_t groups = getGroups();
+
+  if (groups == 0) {
+    return emitOpError("groups must be a positive integer");
+  }
```
Comment (djstefanovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:1592]: Should this also guard groups <=0?

**Review by djstefanovicTT — APPROVED** [review]

File: lib/Dialect/TTIR/IR/TTIROps.cpp (hunk 2 above)
Comment (sonalibaskaran2499) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:1592]: groups is a unit32_t so it can never be negative so already that covers the non positive cases.

**Review by sshonTT — APPROVED** [review]

File: runtime/lib/ttnn/operations/conv/conv1d.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  std::optional<::ttnn::DataType> outputDtype;
+  if (op->output_dtype()) {
+    outputDtype =
+        ::tt::runtime::ttnn::utils::toTTNNDataType(*(op->output_dtype()));
+  }
+
+  ::ttnn::Conv1dConfig conv1dConfig;
+  if (op->conv2d_config()) {
+    conv1dConfig = utils::createConv2dConfig(op->conv2d_config());
```
Comment (jameszianxuTT) [inline runtime/lib/ttnn/operations/conv/conv1d.cpp:49]: not sure about the metal impl, but variable of type conv1dConfig can be assigned to a conv2DConfig? I guess you would probably hit compile failure if this is not true, but this is a bit odd

File: runtime/lib/ttnn/operations/conv/conv1d.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  std::optional<::ttnn::Tensor> bias =
+      op->bias()
+          ? std::make_optional(tensorPool.getTTNNTensorAndValidate(op->bias()))
+          : std::nullopt;
+
+  LOG_ASSERT(op->padding()->size() == 2,
+             "Padding expected to have 2 elements ([pL, pR])");
+
+  // conv1d padding is a (left, right) pair, mapped to the std::array<2>
+  // alternative of the TTNN variant.
```
Comment (jameszianxuTT) [inline runtime/lib/ttnn/operations/conv/conv1d.cpp:35]: Is there any particular reason to block the TTNN conv1d API supporting a single uint32_t for symmetrical padding, though you mention here to only support the array<2> input? 

`padding (int or tuple[int, int]) – Zero-padding added to both sides of the input. pad_length or [pad_left, pad_right].` from the docs

File: runtime/lib/ttnn/operations/conv/conv1d.cpp (hunk 3 above)
Comment (sonalibaskaran2499) [reply runtime/lib/ttnn/operations/conv/conv1d.cpp:49]: In tt-metal Conv1dConfig is a type alias for ttnn::prim::Conv2dConfig (using Conv1dConfig = ttnn::prim::Conv2dConfig; in conv1d.hpp), and createConv2dConfig returns ::ttnn::Conv2dConfig, which is also ttnn::prim::Conv2dConfig. So both sides are the exact same type conv1d is implemented as conv2d with a degenerate dim, so there's no separate config struct to build. Same reasoning for Conv1dSliceConfig.

Comment (jameszianxuTT) [reply runtime/lib/ttnn/operations/conv/conv1d.cpp:49]: interesting, thanks!

File: runtime/lib/ttnn/operations/conv/conv1d.cpp (hunk 4 above)
Comment (sonalibaskaran2499) [reply runtime/lib/ttnn/operations/conv/conv1d.cpp:35]: Symmetric padding is supported, it's normalized to [p, p] during TTIRToTTNN lowering (a scalar or 1-element attr is expanded to {p, p}, see Conv1dOpConversionPattern), so it reaches the runtime as a 2-element [pL, pR] array. The ttnn scalar uint32_t form is just shorthand for {p, p} and carries no extra capability, so we canonicalize to the array<2> alternative rather than plumbing a redundant "scalar vs pair" representation through the flatbuffer.

Comment (jameszianxuTT) [reply runtime/lib/ttnn/operations/conv/conv1d.cpp:35]: I see, thanks! I didn't see that it the padding is canonicalized at an earlier pass. tbh I prefer if the runtime op APIs reflect the TTNN APIs closely, but we don't really have any consumers outside the compiler path, so this is my excess pedantry, sorry!

**Review by jameszianxuTT — APPROVED** [review]

File: runtime/lib/ttnn/operations/conv/conv1d.cpp (hunk 4 above)
Comment (sonalibaskaran2499) [reply runtime/lib/ttnn/operations/conv/conv1d.cpp:35]: No problem! mirroring the TTNN API closely is a reasonable default. Since the only consumer is the compiler path, canonicalizing padding to [pL, pR] upstream keeps the runtime simpler. Thanks for the review!
